### PR TITLE
dev/core#2115 Switch membership type browse page to use v4api for get, thus using acls properly

### PR DESCRIPTION
**Pull request by eileenmcnaughton**
Overview
----------------------------------------
dev/core#2115 Switch membership type browse page to use v4api for get, thus using acls properly -  ie the page at
civicrm/admin/member/membershipType?reset=1

Before
----------------------------------------
Dao retrieval is used and acl work that belongs in an extension is in the core page

After
----------------------------------------
The v4 api is used and the work is done by existing code in financialtypeacls

Technical Details
----------------------------------------
I have only addressed the get call as quite a bit of code change is required for that because

1) we have an array not an object
2) we don't need to covert strings to arrays
3) some variation in the assigned details - as captured in the test https://github.com/civicrm/civicrm-core/pull/20768/files#diff-524ce4dd9256bf434a041487ed7371d32f46b8d8a4826f2e9d135ed8555cbf73L33 - which are basically more strict typing and values are assigned for every row regardless of whether they are in the db (e.g period_rollover_day is in the array even if null)

Comments
----------------------------------------
Note that I added a test in preparation for this PR & this PR alters that test - which demonstrates how this PR changes the output. Per above I think the changes to output are fine/good

### CODE REVIEW

File: CRM/Member/Page/MembershipType.php [hunk 1]
```diff
@@ -84,74 +86,83 @@ public function run() {
...
-      if ($dao->relationship_type_id) {
+      $membershipType[$type['id']]['relationshipTypeName'] = NULL;
+      // Ideally the v4 template would handle the v4 names for these fields - however, that
+      // requires updating edit-in-place so it is a 'todo' for now.
+      $membershipType[$type['id']]['visibility'] = $type['visibility:label'];
+      $membershipType[$type['id']]['period_type'] = $type['period_type:label'];
+      $membershipType[$type['id']]['relationshipTypeName'] = NULL;
+      if ($type['relationship_type_id']) {
         //If membership associated with 2 or more relationship then display all relationship with comma separated
-        $relTypeIds = explode(CRM_Core_DAO::VALUE_SEPARATOR, $dao->relationship_type_id);
```
Comment (eileenmcnaughton) [inline CRM/Member/Page/MembershipType.php:114]: already an array from the api

File: CRM/Member/Page/MembershipType.php [hunk 2]
```diff
@@ -84,74 +86,83 @@ public function run() {
...
+            $membershipType[$type['id']]['relationshipTypeName'] .= ', ';
           }
-          $membershipType[$dao->id]['relationshipTypeName'] .= CRM_Core_DAO::getFieldValue('CRM_Contact_DAO_RelationshipType',
+          $membershipType[$type['id']]['relationshipTypeName'] .= CRM_Core_DAO::getFieldValue('CRM_Contact_DAO_RelationshipType',
             $value, $relationshipName
           );
         }
       }
-      if (CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus() && !CRM_Core_Permission::check('edit contributions of type ' . CRM_Contribute_PseudoConstant::financialType($dao->financial_type_id))) {
+      if (CRM_Financial_BAO_FinancialType::isACLFinancialTypeStatus() && !CRM_Core_Permission::check('edit contributions of type ' . CRM_Contribute_PseudoConstant::financialType($type['financial_type_id']))) {
```
Comment (eileenmcnaughton) [inline CRM/Member/Page/MembershipType.php:138]: Switching these over is out of scope but @colemanw's new api should work

File: tests/phpunit/CRM/Member/Page/MembershipTypeTest.php [hunk 3]
```diff
@@ -30,19 +30,26 @@ public function testMembershipTypePage(): void {
         'id' => '1',
         'domain_id' => '1',
         'name' => 'General',
-        'membership_type' => 'General',
         'member_of_contact_id' => '3',
         'financial_type_id' => '2',
-        'minimum_fee' => '0.000000000',
+        'minimum_fee' => 0.0,
```
Comment (eileenmcnaughton) [inline tests/phpunit/CRM/Member/Page/MembershipTypeTest.php:35]: note this is formatted by smarty

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton Just tested by making `General` type as optional recur, but it is displayed as `Required` on the page -

image

Could be a v4 bug?

**Comment by eileenmcnaughton** [comment]
@colemanw - I just tried to reproduce what @jitendrapurohit identified above - but I hit a new regression that I think relates to your recent changes - I'm seeing

image

even though id is specifically requested in the api call.

Note to test try the page civicrm/admin/member/membershipType?reset=1

```
   $membershipType = (array) MembershipType::get()
      ->addOrderBy('weight')
      ->setSelect([
        'id',
        'domain_id',
        'name',
        'fixed_period_start_day',
        'fixed_period_rollover_day',
        'max_related',
        'relationship_type_id',
        'relationship_direction',
        'member_of_contact_id',
        'financial_type_id',
        'minimum_fee',
        'duration_unit',
        'duration_interval',
        'period_type:label',
        'visibility:label',
        'weight',
        'auto_renew',
        'is_active',
      ])->execute()->indexBy('id');
```

**Comment by colemanw** [comment]
I don't get that at all @eileenmcnaughton - it works fine in the API Explorer:
image

**Comment by eileenmcnaughton** [comment]
@colemanw well pull this down & see if you get the error I did - maybe I was losing the plot?

**Comment by jitendrapurohit** [comment]
@eileenmcnaughton @colemanw I think the problem is with return value of `auto_renew` field.

APIv4 returns a boolean true/false whereas the db stores `0, 1, 2`? 

>0 = No auto-renew option; 1 = Give option, but not required; 2 = Auto-renew required;

image

**Comment by colemanw** [comment]
Well spotted @jitendrapurohit  - try this fix: #20840

**Comment by eileenmcnaughton** [comment]
@colemanw shall we merge this one on the assumption we will figure out the auto_renew in the other one?

**Comment by colemanw** [comment]
If this is causing a regression, I'd rather wait until that one is merged and then re-check.

**Comment by eileenmcnaughton** [comment]
OK - I'm just conscious it cleans up a tonne of smarty notices & it would be easy for someone else to wind up fixing those differently in the interim

**Comment by jitendrapurohit** [comment]
Confirmed that this + https://github.com/civicrm/civicrm-core/pull/20840 fixes the bug mentioned above. Thanks @eileenmcnaughton @colemanw 👍

**Comment by colemanw** [comment]
Ok then this is fine to merge and the other can be merged when tests pass.

**Pull request merged**